### PR TITLE
Filtering, sort and group data

### PR DESCRIPTION
**Pull request by mohseenrm**
## Description

PR to enable filtering, sorting and grouping table data based on problem discussed in loom video.

**Notion**: https://www.notion.so/streak/Mohseen-Mukaddam-810cbb1f89704881a623949e2aa7da62

**Design Doc**: https://docs.google.com/document/d/1XreMNJ2N1XMJ1UV496b43wGYP8dCTLHkZbynyd70Rv0/edit?usp=sharing

Screenshot 2024-05-25 at 10 10 56

**Loom**: https://www.loom.com/share/301de568e7e447ca87b8cf43c5f33de1?sid=cdc7668c-ee6d-42f1-9733-77881e5f6a0d

### CODE REVIEW

File: app/page.tsx [hunk 1]
```diff
@@ -1,113 +1,435 @@
-import Image from "next/image";
+"use client"
+import { useEffect, useCallback, useState, useMemo, useRef } from "react"
+import Title from "@/app/components/title"
+import { Select, Card, Table, Tag } from "antd"
+import type { TableColumnsType } from "antd"
+import TableData from "@/fixtures/data.json"
```
Comment (mohseenrm) [inline app/page.tsx:6]: Using data generated from Faker script for demo

File: app/page.tsx [hunk 2]
```diff
@@ -1,113 +1,435 @@
...
-            Deploy{" "}
-            <span className="inline-block transition-transform group-hover:translate-x-1 motion-reduce:transform-none">
-              -&gt;
-            </span>
-          </h2>
-          <p className="m-0 max-w-[30ch] text-balance text-sm opacity-50">
-            Instantly deploy your Next.js site to a shareable URL with Vercel.
-          </p>
-        </a>
-      </div>
```
Comment (mohseenrm) [inline app/page.tsx:110]: Initial version is code generated from `create-next-app`

**Comment by bkorobeinikov** [comment]
Looks great. The design and color per operation are a nice touch. 

Could you highlight things that you would improve or do differently if there were more time? What considerations did you have while designing the data types and component structure?

**Comment by mohseenrm** [comment]
Oh cool thanks @bkorobeinikov!

> Could you highlight things that you would improve or do differently if there were more time?
Sure yeah, the current filter + group feature works with these operations: >, >=, <, <=, =, !=

It would be nice to support `LIKE` or regex like syntax for querying the data. (caveat with this is, it assumes the data field is of type string, would need a way to apply it semantically across different data types, ex: dates, currency, floats, etc)

> What considerations did you have while designing the data types and component structure?

I worked backwards from the intended goal, of filtering, sorting and grouping data based on fields.

We could apply a list of filters that constrains the source data based on operations, so something like `Array<{ field: TColumn, operation: OperationEnum, value: string | Date | number | boolean }` made sense since we could apply them to the source data and get deterministic results back.

I considered adding `GroupBy` field in to it, but it would be more of a special case operation and would not provide deterministic results based on when the `GroupBy` filter was applied.

For indexing the values of fields, I use a simpler `Set` and later convert it to arrays to ensure we don't have duplicates in autocomplete suggestions.

The components don't contain much of the business logic but simply render based on props. Most of the filtering and grouping logic is moved to `models/TableData.ts` to allow it to be easier to isolate and test v/s bake it into the components

**Comment by bkorobeinikov** [comment]
> It would be nice to support `LIKE` or regex like syntax for querying the data. (caveat with this is, it assumes the data field is of type string, would need a way to apply it semantically across different data types, ex: dates, currency, floats, etc)

could you provide an example of how would you handle this (interface snippet or a description). Different columns could have a different set of operations and each of these operations would have its own value type. For example, Date type column will have operations different from number type columns. Also, Date column could have operations like "is between" which would have `from` and `to` values. or A boolean column that will have `is true` operation that doesn't have any value. 

Also, could you cover the issue with maintability of react components, and how and why would you use composition to approach UI implementation for this usecase.

**Comment by mohseenrm** [comment]
> could you provide an example of how would you handle this (interface snippet or a description).

Sure yeah, good point. I meant `LIKE` / regex applies to string but cannot uniformly be applied to date/currency data types. For adding operations per data type something like this would help.

```typescript
export const Operator = {
  ">": "GREATER_THAN",
  "<": "LESS_THAN",
  "=": "EQUALS",
  ">=": "GREATER_THAN_OR_EQUAL",
  "<=": "LESS_THAN_OR_EQUAL",
  "!=": "NOT_EQUALS",
} as const
type GenericQueryField = {
  field: Field
  operator: Operations
  value: Value
  friendlyValue: string
}
const StringOperator = {
  ...Operator,
  contains: "CONTAINS",
  startsWith: "STARTS_WITH",
  endsWith: "ENDS_WITH",
} as const
type StringQueryField = {
  field: Field
  operator: (typeof StringOperator)[keyof typeof StringOperator]
  value: string
  friendlyValue: string
}
const NumberOperator = {
  ...Operator,
  between: "BETWEEN",
} as const
type NumberQueryField = {
  field: Field
  operator: (typeof NumberOperator)[keyof typeof NumberOperator]
  value: number | { start: number; end: number }
  friendlyValue: string
}
const DateOperator = {
  ...Operator,
  between: "BETWEEN",
} as const
type DateQueryField = {
  field: Field
  operator: (typeof DateOperator)[keyof typeof DateOperator]
  value: Date | { start: Date; end: Date }
  friendlyValue: string
}
export type QueryField =
  | GenericQueryField
  | StringQueryField
  | NumberQueryField
  | DateQueryField

```

**Comment by mohseenrm** [comment]
> Also, could you cover the issue with maintability of react components, and how and why would you use composition to approach UI implementation for this usecase.

Sure yeah, if we had a widget that fetched fields, data, operations from an external source via APIs. It would be beneficial to have a HoC or wrapper component that handled data fetching, querying, cache invalidation, etc.

While the UI components like `QueryBar`, `TableData` continue rendering data based on props. I generally split up components based on business logic, data fetching and rendering logic. In this case the business logic was moved to `models/TableData.ts`

**Pull request merged**